Convert Migrate wizard into a Connection Wizard

MissingRoberto opened the pull request:
## Why?

- Unified experienced connecting to a repository for the users.
- Less code to maintain :)

## How?

- Convert `MigrationToWizard` into a `ConnectWizard` it will establish the connection to a repository offering a migration step if legacy storage is detected.
- Consolidate routes from `/migrate` and `/new` into `/connect`.
- Remove the `NewRepositoryPage`.
- Clean up logic about redirection based on legacy storage in the `ConnectPage`.
- Improve the code to detect if the active step is the last one.

## Screenshots

![Screenshot 2025-03-04 at 19 57 08](https://github.com/user-attachments/assets/41a64037-292e-41a2-9092-7967346f816f)
![Screenshot 2025-03-04 at 19 58 27](https://github.com/user-attachments/assets/7ee7a05a-1a8b-44cb-a92b-5c9c41815101)

MissingRoberto committed:
Convert Migrate wizard into a Connection Wizard

MissingRoberto commented on public/app/features/provisioning/Wizard/ConnectPage.tsx:
```diff
@@ -0,0 +1,15 @@
+import { Page } from 'app/core/components/Page/Page';
+import { ProvisioningWizard } from './ProvisioningWizard';
+
+export default function ConnectPage() {
```
this just got simplified and renamed from `MigrateToProvisioningPage`

ryantxu commented:
With https://github.com/grafana/grafana/pull/101572 -- i don't think this exact form is necessary

BUT, i do think we want to evolve this so the last step could call sync rather than migrate.

We need a wizard that helps when you want connect to a repository that already has things and synchronize -- where we can specify folder vs instance sync targert

MissingRoberto committed:
Merge remote-tracking branch 'origin/grafana-git-ui-sync' into refactor/use-wizard-to-configure-any-repository

MissingRoberto committed:
Remove duplicate empty state

MissingRoberto committed:
Allow users to select target in the first step

MissingRoberto committed:
Remove file created by merge

MissingRoberto committed:
Select target based on existing connections

MissingRoberto committed:
Default option for targets and explainatory alert

MissingRoberto committed:
Do not display connect button if single connection

MissingRoberto committed:
Display target as tag in repository card

MissingRoberto committed:
Add Pull Step

MissingRoberto commented:
Added / clean up logic around single vs folder targets plus added a simple "Pull Dashboards" step which would be alternative to migrate depending in situations I would like to implement in a follow up PR. For now it's legacy vs unified storage. 

- Target Selection: Added functionality to select targets based on existing connections and implemented a default option with an explanatory alert
- Connection UI: Removed the connect button when there's only a single connection
- Repository Card: Added target display as a tag in the repository card.
- Pull Step: Added a new pull step functionality.

![Screenshot 2025-03-05 at 18 14 47](https://github.com/user-attachments/assets/14c3d44a-780b-42f9-92e0-bd56d4610a85)
![Screenshot 2025-03-05 at 18 15 08](https://github.com/user-attachments/assets/51b00a3c-1dc8-487b-ade8-ede5c38decfc)
![Screenshot 2025-03-05 at 18 22 50](https://github.com/user-attachments/assets/b667ae15-dc37-43b5-b609-7dddea140601)
![Screenshot 2025-03-05 at 18 35 36](https://github.com/user-attachments/assets/6278f49a-14b9-4c7d-b3e6-7e36071d1ceb)
![Screenshot 2025-03-05 at 18 48 56](https://github.com/user-attachments/assets/3b2c8125-b151-4509-bf90-074e7b4390c6)


Please, take another look and see how it feels. 


PD: the pull vs migrate we can implement in a follow up :)

ryantxu commented:
🤔  I think we still need to explicit paths
1. Migrate -- move data *from* grafana to repo (and back)
2. Connect -- keep data in grafana, pull it from repo

Landing page:
<img width="782" alt="image" src="https://github.com/user-attachments/assets/b6df1c54-b7e8-402e-a7ff-0b1a799aaa82" />

Migrate wizard... what we have working in the main feature branch
<img width="765" alt="image" src="https://github.com/user-attachments/assets/6c002ff7-edc4-4548-ae2b-baf81b8e4295" />
(calls migrate job at the end)

Connect wizard, 
<img width="793" alt="image" src="https://github.com/user-attachments/assets/afdd5637-c212-4ee3-82bf-f8406ce8dded" />
The link only exists if we are already using unified storage -- legacy storage must migrate

ryantxu commented:
Thinking more... I understand how this flow would work:
<img width="807" alt="image" src="https://github.com/user-attachments/assets/25f4c954-237f-4465-9182-78f0bebcb345" />

I don't yet understand how we can ask instance vs folder as our first question:
<img width="766" alt="image" src="https://github.com/user-attachments/assets/c7748020-358f-45ee-8cea-660585d7a93b" />

MissingRoberto commented:
> Landing page:
Let's exclude that one from this PR. I am working on a separate PR to clean up our components and make things look like Patrick's design: 

![image](https://github.com/user-attachments/assets/961cd902-d2ac-44c3-932a-964074e161b8)

MissingRoberto commented:
I think that the part we are not aligned in that flow chart is it's what's possible or mandatory for folder sync and what determines what. 

Here is how I see it: 


- Existing repositories determine if the user can pick between target or folder. 
- Migrate / Push selection is determined by the target and if the instance is empty.
- Migrate Folder Target doesn't make sense until we support folder selection in migration.

In this way, we offer less options to users during the onboarding process. 

![image](https://github.com/user-attachments/assets/6db6817b-2490-4dc3-86e3-48ffdf3f7317)

Clarity-89 commented:
@MissingRoberto make sure to fix all the FE lint issues you see in CI.

MissingRoberto commented:
@Clarity-89 sure, I think that one is coming from another PR though (probably from me as well)

Clarity-89 commented:
> @Clarity-89 sure, I think that one is coming from another PR though (probably from me as well)

You might need to marge the feature branch into this one, also some of the issues will be fixed by https://github.com/grafana/grafana/pull/101601, so maybe you can wait till that one is merged.

MissingRoberto committed:
Fix linting

Clarity-89 commented on public/app/features/provisioning/Wizard/ConnectionStep.tsx:
```diff
@@ -40,6 +37,22 @@ export function ConnectionStep() {
+  const [instanceConnected, folderConnected] = checkSyncSettings(settingsQuery.data);
+
+  const availableOptions = modeOptions.filter((option) => {
+    if (folderConnected) return option.value === 'folder';
+    if (instanceConnected) return option.value === 'instance';
```
```suggestion
    if (folderConnected) {
      return option.value === 'folder';
    }
    if (instanceConnected) {
      return option.value === 'instance';
    } 
```

Clarity-89 commented on public/app/features/provisioning/Wizard/ConnectionStep.tsx:
```diff
@@ -40,6 +37,22 @@ export function ConnectionStep() {
+  });
+
+  useEffect(() => {
+    if (folderConnected) {
+      setValue('repository.sync.target', 'folder');
+    } else {
+      setValue('repository.sync.target', 'instance');
+    }
```
```suggestion
    if (folderConnected) {
      setValue('repository.sync.target', folderConnected ? 'folder' : 'instance');
    }
```

Clarity-89 commented on public/app/features/provisioning/Wizard/PullStep.tsx:
```diff
@@ -0,0 +1,22 @@
+import { useEffect } from 'react';
```
Do we need a separate step for this or can this info be displayed elsewhere?

MissingRoberto replied:
I created a different step because we are still discussing if we want to trigger the initial sync synchronously as we do with the migrate step. So I was paving the group for @ryantxu to experiment on that. Backend will likely need changes for that.

Clarity-89 commented on public/app/features/provisioning/utils.ts:
```diff
@@ -0,0 +1,10 @@
+  if (!settings?.items?.length) {
+    return [false, false];
+  }
+  const instanceConnected = settings.items.some((item) => item.target === 'instance');
+  const folderConnected = settings.items.some((item) => item.target === 'folder');
```
Aren't these two mutually exclusive, e.g., it can be either instance or folder connected?

MissingRoberto replied:
There is also the third situation in which nothing is connected. Can you guide on what the nicest syntax for that will be?

Clarity-89 replied:
I'd just use separate `isInstance` and `isFolder` functions, but having an empty options complicates that. I guess the current approach is ok then.

MissingRoberto commented:
**offline discussion**: 
We believe that the selection of what the user wants to do it is the most important thing and for that reason it should come first. 

In this case, this is if the user wants to simply connect or migrate the instance. The 

We will have to play a bit between landing page and they way it looks and the wording we use but this is an aspect that we can present and discuss in the UX meeting. 

For this PR, I will add 2 buttons to initiate the stepper for connections and / or migrations.

MissingRoberto committed:
Merge remote-tracking branch 'origin/grafana-git-ui-sync' into refactor/use-wizard-to-configure-any-repository

MissingRoberto committed:
User decides if migrate or connect

MissingRoberto committed:
Improve style based based on review

MissingRoberto commented:
Changed the approach to have 2 buttons available for the user in the landing page to click on migrate or connect. 

In follow-up PRs, we will: 
- Improve the stying of the landing page. 
- Detect if instance or target branch are empty. 
- Initial Sync visible to users.

ryantxu reviewed (approved):
nice we are getting there!
